### PR TITLE
chore(tiering): add protection against overruning memory budget

### PR DESCRIPTION
**Pull request by romange**
Also, add hit/misses stats.

### CODE REVIEW

File: src/server/string_family.cc [hunk 1]
```diff
@@ -532,9 +540,18 @@ struct GetReplies {
...
+                              DbSlice* db_slice) {
+  auto* ts = db_slice->shard_owner()->tiered_storage();
+  if (pv.IsExternal()) {
+    db_slice->IncRamMisses();
+    return StringValue{ts->Read(dbid, key, pv)};
+  }
+
+  db_slice->IncRamHits();
+  if (pv.HasStashPending()) {
+    ts->CancelStash(dbid, key, &pv);
```
Comment (dranikpg) [inline src/server/string_family.cc:552]: this is already done by FindInternal

File: src/server/string_family.cc [hunk 2]
```diff
@@ -448,9 +450,15 @@ SinkReplyBuilder::MGetResponse OpMGet(util::fb2::BlockingCounter wait_bc, bool f
...
       };
+      db_slice.IncRamMisses();
       shard->tiered_storage()->Read(t->GetDbIndex(), it.key(), it->second, std::move(cb));
     } else {
+      db_slice.IncRamHits();
       CopyValueToBuffer(it->second, next);
+      it->first.SetTouched(true);
+      if (it->second.HasStashPending()) {
+        shard->tiered_storage()->CancelStash(index, it.key(), &it->second);
+      }
```
Comment (dranikpg) [inline src/server/string_family.cc:461]: Same here, all that code just makes high level code more difficult to read 😢

File: src/server/tiered_storage.cc [hunk 3]
```diff
@@ -423,11 +430,18 @@ void TieredStorage::RunOffloading(DbIndex dbid) {
...
-    TryStash(dbid, it->first.GetSlice(&tmp), &it->second);
+  string tmp;
+  auto cb = [this, dbid, &tmp](PrimeIterator it) mutable {
+    if (ShouldStash(it->second)) {
+      if (it->first.WasTouched()) {
+        it->first.SetTouched(false);
+      } else {
+        TryStash(dbid, it->first.GetSlice(&tmp), &it->second);
+      }
+    }
```
Comment (dranikpg) [inline src/server/tiered_storage.cc:441]: So we stick with https://github.com/dragonflydb/dragonfly/pull/3069

File: src/server/string_family.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/string_family.cc:461]: And you also won't need to expose changing flags on a const CompactObject

**Review by dranikpg — APPROVED** [review]

File: src/server/string_family.cc (hunk 2 above)
Comment (romange) [reply src/server/string_family.cc:461]: Thanks, I will check again and will add comments here

File: src/server/tiered_storage.cc (hunk 3 above)
Comment (romange) [reply src/server/tiered_storage.cc:441]: I do not think it's exactly the same like what you did. I have a task to read  https://db.in.tum.de/~leis/papers/leanstore.pdf but I think second chance algorithm https://www.geeksforgeeks.org/second-chance-or-clock-page-replacement-policy/ is what we can implement now, and it is a good baseline for the launch. Do you have objections?

File: src/server/string_family.cc (hunk 1 above)
Comment (romange) [reply src/server/string_family.cc:552]: it is not, it is handled in PreUpdate

Comment (romange) [reply src/server/string_family.cc:552]: so should we move it from there to FindInternal?

File: src/server/tiered_storage.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/tiered_storage.cc:441]: No, no objections 😌

**Pull request merged**